velox/expression/SignatureBinder.cpp

```diff
@@ -127,15 +127,34 @@ bool SignatureBinder::tryBind(
   if (it == bindings_.end()) {
     // concrete type
     auto typeName = boost::algorithm::to_upper_copy(typeSignature.baseType());
-    if (isDecimalName(typeName)) {
-      VELOX_USER_FAIL("Use 'DECIMAL' in the signature.");
+
+    if (isDecimalName(typeName) && typeName != actualType->kindName()) {
+      // For Scalar decimal functions, the TypeSignature has specific decimal
+      // type, and the actual type must match.
+      return false;
     }
-    if (isDecimalKind(actualType->kind()) && isCommonDecimalName(typeName)) {
+
+    if (isDecimalKind(actualType->kind()) &&
+        (isDecimalName(typeName) || isCommonDecimalName(typeName))) {
       const auto& variables = typeSignature.variables();
       VELOX_CHECK_EQ(variables.size(), 2);
       const auto& [precision, scale] = getDecimalPrecisionScale(*actualType);
-      variables_.emplace(variables[0], precision);
-      variables_.emplace(variables[1], scale);
+      auto variablesItr = variables_.find(variables[0]);
+      if (variablesItr == variables_.end()) {
+        variables_.emplace(variables[0], precision);
+      } else if (variablesItr->second != precision) {
+        // If precision variable is already initialized but precision values
```

Comment:
nit: A previous instance of the same precision variable does not match in value.